addons/mail/models/mail_followers.py

```diff
@@ -116,39 +116,57 @@ def _get_recipient_data(self, records, subtype_id, pids=None, cids=None):
-    FROM res_partner partner
-    LEFT JOIN res_users users ON users.partner_id = partner.id AND users.active
-    LEFT JOIN res_groups_users_rel groups_rel ON groups_rel.uid = users.id
-    LEFT JOIN res_groups groups ON groups.id = groups_rel.gid
-    WHERE EXISTS (
-        SELECT partner_id FROM sub_followers
-        WHERE sub_followers.channel_id IS NULL
-            AND sub_followers.partner_id = partner.id
-            AND (coalesce(sub_followers.internal, false) <> TRUE OR coalesce(partner.partner_share, false) <> TRUE)
-    ) %s
-    GROUP BY partner.id, users.notification_type
-UNION
-SELECT NULL AS pid, channel.id AS cid,
-        TRUE as active, NULL AS pshare, channel.channel_type AS ctype,
-        CASE WHEN channel.email_send = TRUE THEN 'email' ELSE 'inbox' END AS notif, NULL AS groups
-    FROM mail_channel channel
-    WHERE EXISTS (
-        SELECT channel_id FROM sub_followers WHERE partner_id IS NULL AND sub_followers.channel_id = channel.id
-    ) %s
-""" % ('OR partner.id IN %s' if pids else '', 'OR channel.id IN %s' if cids else '')
+SELECT DISTINCT ON(pid, cid) * FROM (
+    WITH sub_followers AS (
+        SELECT fol.id, fol.partner_id, fol.channel_id, subtype.internal
+        FROM mail_followers fol
+            RIGHT JOIN mail_followers_mail_message_subtype_rel subrel
+            ON subrel.mail_followers_id = fol.id
+            RIGHT JOIN mail_message_subtype subtype
+            ON subtype.id = subrel.mail_message_subtype_id
+        WHERE subrel.mail_message_subtype_id = %%s AND fol.res_model = %%s AND fol.res_id IN %%s
+    )
+    %s
+    SELECT partner.id as pid, NULL AS cid,
+            partner.active as active, partner.partner_share as pshare, NULL as ctype,
+            users.notification_type AS notif, array_agg(groups.id) AS groups
+        FROM res_partner partner
+        LEFT JOIN res_users users ON users.partner_id = partner.id AND users.active
+        LEFT JOIN res_groups_users_rel groups_rel ON groups_rel.uid = users.id
+        LEFT JOIN res_groups groups ON groups.id = groups_rel.gid
+        INNER JOIN sub_followers subfol ON subfol.partner_id = partner.id
+
+            WHERE
+            subfol.channel_id IS NULL
+            AND (
+                coalesce(subfol.internal, false) <> TRUE
+                OR coalesce(partner.partner_share, false) <> TRUE
+            )
+        GROUP BY partner.id, users.notification_type
+    UNION ALL
+    SELECT NULL AS pid, channel.id AS cid,
+            TRUE as active, NULL AS pshare, channel.channel_type AS ctype,
+            CASE WHEN channel.email_send = TRUE THEN 'email' ELSE 'inbox' END AS notif, NULL AS groups
+        FROM mail_channel channel
+        WHERE EXISTS (
+            SELECT channel_id FROM sub_followers WHERE partner_id IS NULL AND sub_followers.channel_id = channel.id
+        ) %s
+) AS x
+ORDER BY pid, cid, notif
+""" % (query_partners if pids else '', 'OR channel.id IN %s' if cids else '')
```

Comment:
To avoid a dynamic query building, you can always add the query_partners in your query. It's just empty when there's no id with clause id=ANY(%s) https://github.com/odoo/odoo/pull/69577/files#diff-38b9c9b6cee861712a501fc816975996e4ed661685b94cb6e24e3dbaeebd7b77R122.

With the tests I've done, not on the same db nor version, I admit, adding clause like `OR channel.id IN %s` makes the query to be slow. Maybe you can `union all` those channels.

Reply:
About channels.

I assume that mail.channel table is not big. And because this is a PR to v12, I'd prefer to keep same code as in v13.

About partners.

If we can we can avoid adding SELECT ... FROM ... JOIN ... LEFT JOIN ... UNION ALL, it's better to don't add it, no?